cmd/osbuild-image-tests/main_test.go

```diff
@@ -189,39 +200,111 @@ func testSSH(t *testing.T, ns netNS) {
 	t.Errorf("ssh test failure, %d attempts were made", attempts)
 }
 
+func testBootUsingQemu(t *testing.T, imagePath string) {
+	err := withNetworkNamespace(func(ns netNS) error {
+		return withBootedQemuImage(imagePath, ns, func() error {
+			testSSH(t, "localhost", testPaths.privateKey, &ns)
+			return nil
+		})
+	})
+	require.NoError(t, err)
+}
+
+func testBootUsingNspawnImage(t *testing.T, imagePath string, outputID string) {
+	err := withNetworkNamespace(func(ns netNS) error {
+		return withBootedNspawnImage(imagePath, outputID, ns, func() error {
+			testSSH(t, "localhost", testPaths.privateKey, &ns)
+			return nil
+		})
+	})
+	require.NoError(t, err)
+}
+
+func testBootUsingNspawnDirectory(t *testing.T, imagePath string, outputID string) {
+	err := withNetworkNamespace(func(ns netNS) error {
+		return withExtractedTarArchive(imagePath, func(dir string) error {
+			return withBootedNspawnDirectory(dir, outputID, ns, func() error {
+				testSSH(t, "localhost", testPaths.privateKey, &ns)
+				return nil
+			})
+		})
+	})
+	require.NoError(t, err)
+}
+
+func testBootUsingAWS(t *testing.T, imagePath string) {
+	creds, err := getAWSCredentialsFromEnv()
+	require.NoError(t, err)
+
+	// if no credentials are given, fall back to qemu
+	if creds == nil {
+		log.Print("no AWS credentials given, falling back to booting using qemu")
+		testBootUsingQemu(t, imagePath)
+		return
+
+	}
+
+	imageName, err := generateRandomString("osbuild-image-tests-image-")
+	require.NoError(t, err)
+
+	e, err := newEC2(creds)
+	require.NoError(t, err)
+
+	// the following line should be done by osbuild-composer at some point
+	err = uploadImageToAWS(creds, imagePath, imageName)
+	assert.NoError(t, err) // do not immediately end the test, try to do some cleanup
+
+	imageDesc, err := describeEC2Image(e, imageName)
+	require.NoError(t, err) // there's nothing we can do at this point
+
+	// delete the image when this function returns
+	defer func() {
+		err := deleteEC2Image(e, imageDesc)
+		if err != nil {
+			log.Printf("cannot delete the ec2 image, resources could have been leaked: %#v", err)
+		}
+	}()
+
+	// If the test is failed at this point, there's no point in continuing.
+	// The deferred cleanup will be called.
+	if t.Failed() {
+		t.FailNow()
+	}
```

Comment:
If you change line 255 to use require instead of assert won't this condition become obsolete ?

Reply:
I want to register the deferred function on line 261 and I couldn't find a cleaner way to write it, therefore I chose this a bit cumbersome, but reliable way.

Reply:
I think you should register the deferred function on line 258 before `require.NoError` b/c in case of error that will short-cirtcuit the test and you won'd perform cleanup, right ? Then this if condition will not be necessary.

Reply:
I changed my mind, the error handling was very confusing and over-engineered, I pushed a new much simpler version, which should do basically the same.